### ISSUE TITLE
nn.LayerNorm breaks gradient flow under specific situations

### ISSUE THREAD

**Issue by rodrigo-castellon** [pytorch/pytorch#63541]
## 🐛 Bug

If you do some computations followed by an nn.LayerNorm followed by .sum().backward(), all gradients become zeroed out, but if you add an additional step before .sum().backward(), such as multiplying by the scalar 0.99, all of the gradients flow perfectly fine. Note that this is a symptom. There are many other bizarre behaviors that stem from the same underlying bug, and I'm not sure what/where the bug is.

## To Reproduce

See the colab notebook below, which demonstrates this bug.
https://colab.research.google.com/drive/1wv6zkZjZ11OIUQm6wahTKMAUjUg4XuKx?usp=sharing

A short code snippet that demonstrates this is
```
import torch
import torch.nn as nn

hidden_dim = 512
layernorm = nn.LayerNorm(hidden_dim)

x = torch.rand(10, hidden_dim, requires_grad=True)

out = layernorm(x)
out.sum().backward()

print(x.grad)
print(x.grad.abs().max().item())

# same as above, except we add a scalar multiplier to the computation graph
x = torch.rand(10, hidden_dim, requires_grad=True)

out = layernorm(x)
(0.99 * out).sum().backward()

print(x.grad)
print(x.grad.abs().max().item())
```

The above prints out something like
```
tensor([[0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.],
        ...,
        [0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.]])
0.0
tensor([[ 2.3842e-06, -3.5763e-06,  3.0994e-06,  ..., -8.3447e-06,
          4.7684e-06, -8.1062e-06],
        [-9.2983e-06, -9.0599e-06, -5.7220e-06,  ...,  1.4305e-06,
          2.8610e-06, -8.8215e-06],
        [-4.5300e-06, -2.3842e-06, -7.6294e-06,  ..., -5.4836e-06,
         -5.7220e-06, -2.8610e-06],
        ...,
        [-7.8678e-06,  3.5763e-06, -4.7684e-06,  ..., -3.8147e-06,
          1.4305e-06, -7.8678e-06],
        [-8.1062e-06, -7.3910e-06, -1.0967e-05,  ..., -1.4305e-06,
          2.3842e-07, -1.0252e-05],
        [ 2.3842e-07, -4.0531e-06, -5.9605e-06,  ...,  4.5300e-06,
          4.7684e-06, -4.7684e-06]])
1.1205673217773438e-05

```

## Expected behavior

We expect gradient flow to not depend on scalar multipliers (as long as the multiplier is not 0). Furthermore, we expect gradients to be nonzero; after all, nn.LayerNorm is a key component in many model architectures.

## Environment

We can run this in Colab today (August 18th, 2021) and see this behavior. Below is the output from the environment collection script from the Colab notebook (unsure if this is really necessary).

```
PyTorch version: 1.9.0+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: version 3.12.0
Libc version: glibc-2.26

Python version: 3.7.11 (default, Jul  3 2021, 18:01:19)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.104+-x86_64-with-Ubuntu-18.04-bionic
Is CUDA available: False
CUDA runtime version: 11.0.221
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.4
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.5
[pip3] torch==1.9.0+cu102
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.10.0
[pip3] torchvision==0.10.0+cu102
[conda] Could not collect
```

## Additional context

Ran into this issue when adapting `nn.TransformerDecoder` to my problem and was checking that the gradients looked ok by calling .sum().backward() on the outputs, but since `nn.TransformerDecoder` ends in a LayerNorm, I saw all zero gradients. Slowed down my project quite a bit to hunt down this bug. Also note that there are many other strange behaviors as a result of the same underlying bug, such as the fact that this only occurs on inputs with the last dimension size being a power of 2 (see Colab notebook for more details). Let me know if I happen to be overlooking something...

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @mruberry @jbschlosser

**Comment by soulitzer** [comment]
It is expected that the gradients are zero since normalization makes it so that the average value of the elements is zero (which makes the `.sum()` always zero as well). The gradient you are seeing when you have a scalar multiple is probably due to numerical precision issues (try different values of hidden_dim and use `torch.double`).

**Comment by rodrigo-castellon** [comment]
Got it, that makes sense.
